### PR TITLE
R json output

### PR DESCRIPTION
**Pull request by robinpaul85**
## Description

Output from hclust.R is returned as JSON. fastclust.R has been renamed to hclust.R. 
Please test this branch with high-sample and high-gene GDC example to ensure that the jsonlite package in R is able to stream data back to nodejs for such cases. 

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [ ] Tests: added and/or passed unit and integration tests, or N/A
- [ ] Todos: commented or documented, or N/A
- [ ] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

File: server/src/mds3.load.js [hunk 1]
```diff
@@ -479,77 +479,41 @@ async function geneExpressionClustering(data, q, ds) {
...
 	const Rinputfile = path.join(serverconfig.cachedir, Math.random().toString() + '.json')
 	await write_file(Rinputfile, JSON.stringify(inputData))
-	const Routput = await lines2R(path.join(serverconfig.binpath, 'utils/fastclust.R'), [], [Rinputfile])
+	const Routput = JSON.parse(await lines2R(path.join(serverconfig.binpath, 'utils/hclust.R'), [], [Rinputfile]))
 	fs.unlink(Rinputfile, () => {})
-	//const r_output_lines = Routput.trim().split('\n')
-	//console.log('r_output_lines:', r_output_lines)
 
 	let row_coordinates = []
+	for (const item of JSON.parse(Routput['RowNodeJson'])) {
```
Comment (xzhou82) [inline server/src/mds3.load.js:486]: should only JSON.parse() once at line 482. e.g. to get rid of this json.parse, delete `toJSON()` in line 105 hclust.r

File: server/utils/hclust.R [hunk 2]
```diff
@@ -123,11 +142,35 @@ SortedColumnNames <- colnames(normalized_matrix)[ColumnDend$order]
...
+})}
+output_df$SortedColumnNames <- {lapply(1:length(SortedColumnNames), function(y){
+    list(sample=SortedColumnNames[y])
+})}
+output_df$OutputMatrix <- {lapply(1:length(normalized_matrix), function(y){
+    list(elem=normalized_matrix[y])
+})}
+#print ("output_json")
+output_json <- toJSON(output_df)
+print (output_json)
```
Comment (gavrielm) [inline server/utils/hclust.R:167]: `toJSON()` will output to standard out. No need for the `print()` statement.

**Review by josephpowi — COMMENTED** [review]
it seems that the 'toJSON' function won't work for streaming data. It might be better to use 'streaamR' package for streaming data to node js if we deal with the big size data.

**Comment by gavrielm** [comment]
What data can be used for testing?

**Comment by xzhou82** [comment]
use non-gdc links in `clustering` section

File: server/src/mds3.load.js [hunk 3]
```diff
@@ -479,77 +479,43 @@ async function geneExpressionClustering(data, q, ds) {
 
 	const Rinputfile = path.join(serverconfig.cachedir, Math.random().toString() + '.json')
 	await write_file(Rinputfile, JSON.stringify(inputData))
-	const Routput = await lines2R(path.join(serverconfig.binpath, 'utils/fastclust.R'), [], [Rinputfile])
+	const Routput = JSON.parse(await lines2R(path.join(serverconfig.binpath, 'utils/hclust.R'), [], [Rinputfile]))
 	fs.unlink(Rinputfile, () => {})
-	//const r_output_lines = Routput.trim().split('\n')
-	//console.log('r_output_lines:', r_output_lines)
 
-	let row_coordinates = []
```
Comment (xzhou82) [inline server/src/mds3.load.js:487]: should not use `let`, use const. the variable is not mutable

**Comment by robinpaul85** [comment]
> it seems that the 'toJSON' function won't work for streaming data. It might be better to use 'streamR' package for streaming data to nodejs if we deal with the big size data.

Yes, that is what I am also a bit afraid. That is why until now, I was refraining from using toJSON()

File: server/utils/hclust.R (hunk 2 above)
Comment (josephpowi) [reply server/utils/hclust.R:162]: I think you need to remove output_json as well when you comment out the print statement. Just "toJSON(output_df)" will work.

**Review by xzhou82 — APPROVED** [review]

**Pull request merged**